### PR TITLE
Add validation

### PR DESCRIPTION
**Pull request by mattkrick**
*Welcome*
it
- [ ] Makes you enter a user name on step 1
- [ ] makes you enter a team name on step 2
- [ ] tells you where in the list of email addresses you messed up on step 3
- [ ] makes you enter a task name longer than 1 character if you decide to give a team member a task on step 3
- [ ] filters duplicate email addresses while parsing on step 3
- [ ] if you try entering an email that already exists in the invitee list, that email will flash bold & pink to let you know that it's already there
- [ ] lets you bypass step 3 if you don't want to invite anyone
- [ ] performs schema validation on the server so even if you're sneaky you can't put poo in our DB
- [ ] applies step1 validation logic (preferredName) to the user settings page 
- [ ] shows schema only after you hit submit or blur the field
- [ ] after submitting a form (eg `inviteesRaw` in step 3) it will _not_ show an error until you attempt to submit or re-blur
- [ ] reduces time delay from step 2 -> step 3 and step 3 -> dashboard
- [ ] rejects bad/duplicate invitations from the team settings page

*Dash*
it
- [ ] trims spaces from projects & actions
- [ ] validates list of invitees when adding a new team
- [ ] validates team name when changing it in team settings
- [ ] fixes a regression in the endMeeting mutation
- [ ] fixes https://github.com/ParabolInc/action-devops/issues/1
- [ ] refactors logout to signout

### CODE REVIEW

**Comment by mattkrick** [comment]
ready for :eyes:
To customize the error messages, check out `/universal/validation`. Make sure they aren't too snarky :smile:

**Comment by jordanh** [comment]
Test cycle:

*Welcome*
it
- [ ] Makes you enter a user name on step 1
- [ ] makes you enter a team name on step 2
- [ ] tells you where in the list of email addresses you messed up on step 3
- [ ] makes you enter a task name longer than 1 character if you decide to give a team member a task on step 3
- [ ] filters duplicate email addresses while parsing on step 3
- [ ] if you try entering an email that already exists in the invitee list, that email will flash bold & pink to let you know that it's already there
- [ ] lets you bypass step 3 if you don't want to invite anyone
- [ ] performs schema validation on the server so even if you're sneaky you can't put poo in our DB
- [ ] applies step1 validation logic (preferredName) to the user settings page 
- [ ] shows schema only after you hit submit or blur the field
- [ ] after submitting a form (eg `inviteesRaw` in step 3) it will _not_ show an error until you attempt to submit or re-blur
- [ ] reduces time delay from step 2 -> step 3 and step 3 -> dashboard
- [ ] rejects bad/duplicate invitations from the team settings page

*Dash*
it
- [ ] trims spaces from projects & actions
- [ ] validates list of invitees when adding a new team
- [ ] validates team name when changing it in team settings
- [ ] fixes a regression in the endMeeting mutation
- [ ] fixes https://github.com/ParabolInc/action-devops/issues/1
- [ ] refactors logout to signout

File: src/universal/components/IconButton/IconButton.js [hunk 1]
```diff
@@ -10,9 +10,9 @@ const IconButton = (props) => {
     styles.base,
     disabled && styles.disabled
   );
-
+  // must declare type="button" or it gets treated as a submit in the welcome wizard. wtf react
```
Comment (jordanh) [inline src/universal/components/IconButton/IconButton.js]: heh. I saw you open this issue. Cray-Z.

File: src/universal/validation/legitify.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
+class Legitity {
```
Comment (jordanh) [inline src/universal/validation/legitify.js]: `Legitity` (-ity) or `Legitify` (-ify)?

File: src/universal/validation/legitify.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        if (Object.keys(res.errors).length > 0) {
+          errors[key] = res.errors;
+        }
+      }
+    }
+  }
+  return {errors, data};
+};
+
+export default legitify;
```
Comment (jordanh) [inline src/universal/validation/legitify.js]: Keep this here or make it its own module?

Comment (jordanh) [inline src/universal/validation/legitify.js]: This is elegant. Monads FTW. Would make a great post!

File: src/universal/validation/templates.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      }
+      return 'That first email doesn\'t look right';
+    }
+    return undefined;
+  });
+
+export const teamName = (value) => value
+  .trim()
+  .required('"The nameless wonder" is better than nothing')
+  .min(2, 'The "A Team" had a longer name than that')
```
Comment (jordanh) [inline src/universal/validation/templates.js]: heh. Safe to try.

**Comment by jordanh** [comment]
Server validation works. Cool! Tested by:

image

**Comment by jordanh** [comment]
I am seeing the schema validation applied onChange, not only onBlur or onSubmit. For example, when updated User Settings preferredName—is that intended?

**Comment by jordanh** [comment]
Opened #558

**Comment by jordanh** [comment]
# Test Cycle 1 Complete!

File: src/universal/validation/legitify.js (hunk 2 above)
Comment (mattkrick) [inline src/universal/validation/legitify.js]: A Legitity is kinda like an Identity monad. but legit.

File: src/universal/validation/legitify.js (hunk 3 above)
Comment (mattkrick) [inline src/universal/validation/legitify.js]: i'll pull it out in a subsequent PR. I feel like it's still a little young to be on its own.

**Comment by mattkrick** [comment]
ugh, redux-form doesn't make it easy to do onBlur validation. instead, they recommend to do onChange validation (default) & then hide the error until the field gets blurred (touched == true). What we could do is write a custom `shouldValidate` function that only calls `validate` when `submitting` is true, but in able to use `submitting` our `onSubmit` function would have to return a promise. probably my biggest gripe with redux-form right now.

**Comment by jordanh** [comment]
I am ok to start with this onChange. I think there is a net-gain to the experience as-is. What do you think?

**Comment by mattkrick** [comment]
yeah, I think that's OK for now.

**Comment by jordanh** [comment]
Fix #558 and let's merge?

**Pull request merged**